[4][Media] Renaming files when containing 100% non-latin ascii

PhilETaylor opened the pull request:
Fixes https://github.com/joomla/joomla-cms/issues/33452

### Summary of Changes

If all that is left after making the file name safe is the extension, then error and say why. 

Also fixes a double slash issue `//` when a filename has been made save. 

### Testing Instructions

WIHTOUT iconv and intl PHP extensions, attempt to rename a FILE to `МИ.png`

### Actual result BEFORE applying this Pull Request

<img width="806" alt="Screenshot 2021-04-30 at 23 13 11" src="https://user-images.githubusercontent.com/400092/116759188-a84d5500-aa09-11eb-8e35-cc798196df3a.png">

and the moved file is created at `/images/png`

### Expected result AFTER applying this Pull Request

<img width="1256" alt="Screenshot 2021-04-30 at 23 12 35" src="https://user-images.githubusercontent.com/400092/116759144-910e6780-aa09-11eb-9566-0824f25e624a.png">


### Documentation Changes Required

Commit:
fix renaming files when containing 100% non-latin ascii

Kostelano commented:
I have tested this item :white_check_mark: successfully on 6dd7e1cddb1183542aeb5b651c7e19c212ac69a9<br /><br />Thank you!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33454">issues.joomla.org/tracker/joomla-cms/33454</a>.</sub>

Fedik commented:
ideally it should be transliterated, but an exception also enough for quick fix :)

PhilETaylor commented:
It can't be if you are missing the PHP extensions to actually do that!

Fedik commented:
> It can't be if you are missing the PHP extensions to actually do that!

It is still possible, we actually doing it for an alias field, 
However it does not cover everything, mostly only for installed languages that provide transliteration in their localise.php

https://github.com/joomla/joomla-cms/blob/e44a1d22202030f8e69ba578e972a152d52f0d62/libraries/src/Filter/OutputFilter.php#L85-L89

and 

https://github.com/joomla/joomla-cms/blob/3a5984d03366f707ef842ee48051af194f68848f/libraries/src/Language/Language.php#L396-L401

richard67 commented on plugins/filesystem/local/src/Adapter/LocalAdapter.php:
```diff
@@ -546,10 +546,16 @@ public function move(string $sourcePath, string $destinationPath, bool $force =
 		$name     = $this->getFileName($destinationPath);
 		$safeName = $this->getSafeName($name);
 
+		// If transliterating could not happen, and 100% of the chars are filtered out, then throw an error. 3 = file extension
```
```suggestion
		// If transliterating could not happen, and all characters except of the file extension are filtered out, then throw an error.
```

infograf768 commented:
for j4, the priority for the alias as well as for upload image, is:

1 intl extension
2 language pack custom transliteration.
3 for alias only, if no ascii, the date of saving

Commit:
Update plugins/filesystem/local/src/Adapter/LocalAdapter.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

PhilETaylor commented:
Well if that case the CMS should be cleaning up the file name before asking the `File` library to write the file. 

The CMS File library is not the place to be injecting lots of language specific code. 

`File::makeSafe` is not the place for language pack transliteration. 

Replacing a failed `makeSafe` with todays date would be worse in my opinion - someone attempting to rename a file, only for that file to not be renamed to what they asked, but to a date... is likely to confuse.

Fedik commented:
> File::makeSafe is not the place for language pack transliteration.

Correct.

Here we have `LocalAdapter::getSafeName()`.
https://github.com/joomla/joomla-cms/blob/ef05728f055127f77c9a2a271b68b32f6f1ce95e/plugins/filesystem/local/src/Adapter/LocalAdapter.php#L777-L779

Additionally, the same issue may affect all methods in `LocalAdapter`, not only `move()`:
```
createFolder()
createFile()
copy()
```

PhilETaylor commented:
` $lang->transliterate('МИМИМИМИ')` also returns a blank string...

Commit:
throw exception if cannot make media name safe

Commit:
Merge branch 'media/renamefilenonascii' of github.com:PhilETaylor/joomla-cms into media/renamefilenonascii

PhilETaylor commented:
I have added the exceptions in the LocalAdapter makeSafe, and added language transliterate with another check as both `File::makeSafe` and $lang->transliterate('МИМИМИМИ') can return a blank string.

Commit:
rename language string

Fedik commented on plugins/filesystem/local/src/Adapter/LocalAdapter.php:
```diff
@@ -772,7 +778,15 @@ private function getEncodedPath(string $path): string
+			throw new \Exception(Text::_('COM_MEDIA_ERROR_MAKESAFE'));
+		}
+
+		if (!$name = Factory::getLanguage()->transliterate($name))
```
After some rest, and look at it again,
I think it will not work as I expected, because `File::makeSafe` already return "empty" name, and so the transliterate will transliterate emptinesssss.

Sorry for confusing.

The first exception after `File::makeSafe` is good here.

PhilETaylor replied:
removed

Commit:
revert second check

richard67 committed:
Merge branch '4.0-dev' into media/renamefilenonascii

richard67 commented:
I've updated the branch to latest 4.0-dev to get rid of the unrelated javascript-cs error in drone.

@Kostelano Could you redo your test? The PR has meanwhile received some change beside that branch update. Thanks in advance.

Kostelano commented:
So, it works a little differently than I expect.

I just take a file with a Latin name and try to ADD Russian letters to its name. I receive an error NOT declared in the PR, but already another one.

![Screenshot_1](https://user-images.githubusercontent.com/8440661/116851111-8109aa00-abfa-11eb-9437-60ab28b76401.png)

If I try to rename the file to a Russian name WITHOUT a combination with Latin names, I will already get the declared error.

![Screenshot_2](https://user-images.githubusercontent.com/8440661/116851275-c62ddc00-abfa-11eb-9ae3-0944d39afe37.png)

It seems to me that both cases should somehow correspond to the same error.

Also, note that errors are duplicated on the same page, similar to the error reported at https://github.com/joomla/joomla-cms/issues/33450.

PhilETaylor commented:
If a file exists with just the latin chars, then that is correct, as makeSafe will remove the russian and leave the other chars (if your server doesn't have iconv/intl installed)

Fedik commented:
I have tested this item :white_check_mark: successfully on cee796056c5cd3034832fa0a030e1936e7ed8375<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33454">issues.joomla.org/tracker/joomla-cms/33454</a>.</sub>

Kostelano commented:
I have tested this item :white_check_mark: successfully on cee796056c5cd3034832fa0a030e1936e7ed8375<br /><br />Ok, the main issue is solved by this PR. Thank you!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33454">issues.joomla.org/tracker/joomla-cms/33454</a>.</sub>

alikon commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33454">issues.joomla.org/tracker/joomla-cms/33454</a>.</sub>

wilsonge commented:
Thanks!
